### PR TITLE
Displayscaling

### PR DESCRIPTION
**Pull request by pakfront**
DRAFT
Use shared methods across MM, MHQ and MML for display scaling and splash screens.
These two requests require this change: 
https://github.com/MegaMek/megameklab/pull/1053
https://github.com/MegaMek/mekhq/pull/3150

### CODE REVIEW

File: megamek/src/megamek/client/ui/swing/util/UIUtil.java [hunk 1]
```diff
@@ -409,7 +411,169 @@ public static void scaleMenu(final JComponent popup) {
...
+
+    /**
+     *
+     * @param currentMonitor
+     * @return The height of the screen taking into account display scaling
+     */
+    public static Dimension getScaledScreenSize(DisplayMode currentMonitor) {
+        int monitorH = currentMonitor.getHeight();
+        int monitorW = currentMonitor.getWidth();
+        int pixelPerInch= Toolkit.getDefaultToolkit().getScreenResolution();
```
Comment (SJuliez) [inline megamek/src/megamek/client/ui/swing/util/UIUtil.java:455]:
```suggestion
        int pixelPerInch = Toolkit.getDefaultToolkit().getScreenResolution();

```

File: megamek/src/megamek/client/ui/swing/util/UIUtil.java [hunk 2]
```diff
@@ -409,7 +411,169 @@ public static void scaleMenu(final JComponent popup) {
...
+        return new Dimension(
+                DEFAULT_DISPLAY_PPI * monitorW / pixelPerInch,
+                DEFAULT_DISPLAY_PPI * monitorH / pixelPerInch);
+    }
+
+    /**
+     *
+     * @param image
+     * @param maxWidth
+     * @param maxHeight
```
Comment (SJuliez) [inline megamek/src/megamek/client/ui/swing/util/UIUtil.java:465]: If the "return " says it all, I suggest removing the other @param

File: megamek/src/megamek/client/ui/swing/util/UIUtil.java [hunk 3]
```diff
@@ -409,7 +411,169 @@ public static void scaleMenu(final JComponent popup) {
...
+     * @return an image with the same aspect ratio that fits within the given bounds, or the existing image if it already does
+     */
+    public static Image constrainImageSize(Image image, ImageObserver observer, int maxWidth, int maxHeight) {
+        Image newImage = image;
+        if (newImage.getWidth(observer) > maxWidth) {
+            newImage = newImage.getScaledInstance(maxWidth, -1, Image.SCALE_DEFAULT);
```
Comment (SJuliez) [inline megamek/src/megamek/client/ui/swing/util/UIUtil.java:471]: Im not sure but I think SCALE_SMOOTH or something like that is better than SCALE_DEFAULT. (And with images that have transparency even that isnt good. I think the best image scaler is used for displaying the scaled image icons)

Comment (SJuliez) [reply megamek/src/megamek/client/ui/swing/util/UIUtil.java:471]: That would be ImageUtil.getscaledInstance

File: megamek/src/megamek/client/ui/swing/util/UIUtil.java [hunk 4]
```diff
@@ -409,7 +411,169 @@ public static void scaleMenu(final JComponent popup) {
...
+    /**
+     *
+     * @param imgSplashFile path to an image on disk
+     * @param parent component
+     * @return a JLabel setup to the correct size to act as a splash screen
+     */
+    public static JLabel createSplashComponent(String imgSplashFile, Component parent) {
+        // Use the current monitor so we don't "overflow" computers whose primary
+        // displays aren't as large as their secondary displays.
+        Dimension scaledMonitorSize = getScaledScreenSize( parent.getGraphicsConfiguration().getDevice().getDisplayMode());
```
Comment (SJuliez) [inline megamek/src/megamek/client/ui/swing/util/UIUtil.java:512]:
```suggestion
        Dimension scaledMonitorSize = getScaledScreenSize(parent.getGraphicsConfiguration().getDevice().getDisplayMode());

```

File: megamek/src/megamek/client/ui/swing/util/UIUtil.java [hunk 5]
```diff
@@ -409,7 +411,169 @@ public static void scaleMenu(final JComponent popup) {
...
+     *
+     * @param imgSplash an image
+     * @param observer
+     * @param scaledMonitorSize the dimensions of the monitor taking into account display scaling
+     * @return a JLabel setup to the correct size to act as a splash screen
+     */
+    public static JLabel createSplashComponent(Image imgSplash, ImageObserver observer, Dimension scaledMonitorSize) {
+        JLabel splash;
+        Dimension maxSize = new Dimension(
+                (int) (scaledMonitorSize.width*0.75),
```
Comment (SJuliez) [inline megamek/src/megamek/client/ui/swing/util/UIUtil.java:538]:
```suggestion
                (int) (scaledMonitorSize.width * 0.75),

```

File: megamek/src/megamek/client/ui/swing/util/UIUtil.java [hunk 6]
```diff
@@ -409,7 +411,169 @@ public static void scaleMenu(final JComponent popup) {
...
+                (int) (scaledMonitorSize.height * 0.75));
+
+        if (imgSplash != null) {
+            imgSplash = UIUtil.constrainImageSize(imgSplash, null, maxSize.width, maxSize.height );
```
Comment (SJuliez) [inline megamek/src/megamek/client/ui/swing/util/UIUtil.java:542]:
```suggestion
            imgSplash = UIUtil.constrainImageSize(imgSplash, null, maxSize.width, maxSize.height);

```

File: megamek/src/megamek/client/ui/swing/util/UIUtil.java [hunk 7]
```diff
@@ -409,7 +411,169 @@ public static void scaleMenu(final JComponent popup) {
...
+        Dimension size = component.getSize();
+        Rectangle r = new Rectangle(scaledScreenSize);
+
+        // center and size if out of bounds
+        if ( (pos.x < 0) || (pos.y < 0) ||
+                (pos.x + size.width > scaledScreenSize.width) ||
+                (pos.y + size.height > scaledScreenSize.getHeight())) {
+            component.setLocationRelativeTo(null);
+        }
+    }
```
Comment (SJuliez) [inline megamek/src/megamek/client/ui/swing/util/UIUtil.java:576]: Seeing that this is quite a bit of code and has no direct connection to the other stuff in UIUtil I'm tempted to suggest moving this to its own "MonitorUtil" or something like that. But I know that Windchild suggested UIUtil and there's nothing wrong with that either. I'll leave it to you. We probably could use some more utility methods for handling multi monitor setups as we often have problems with these.

File: megamek/src/megamek/client/ui/swing/util/UIUtil.java [hunk 8]
```diff
@@ -409,7 +411,169 @@ public static void scaleMenu(final JComponent popup) {
...
+
+    /**
+     *
+     * @param currentMonitor
+     * @return The height of the screen taking into account display scaling
+     */
+    public static int getScaledScreenHeight(DisplayMode currentMonitor) {
+        int monitorH = currentMonitor.getHeight();
+        int pixelPerInch= Toolkit.getDefaultToolkit().getScreenResolution();
```
Comment (SJuliez) [inline megamek/src/megamek/client/ui/swing/util/UIUtil.java:434]:
```suggestion
        int pixelPerInch = Toolkit.getDefaultToolkit().getScreenResolution();

```

File: megamek/src/megamek/client/ui/swing/util/UIUtil.java [hunk 9]
```diff
@@ -409,7 +411,169 @@ public static void scaleMenu(final JComponent popup) {
...
+
+
+    /**
+     *
+     * @param currentMonitor
+     * @return the width of the screen taking into account display scaling
+     */
+    public static int getScaledScreenWidth(DisplayMode currentMonitor) {
+        int monitorW = currentMonitor.getWidth();
+        int pixelPerInch= Toolkit.getDefaultToolkit().getScreenResolution();
```
Comment (SJuliez) [inline megamek/src/megamek/client/ui/swing/util/UIUtil.java:423]:
```suggestion
        int pixelPerInch = Toolkit.getDefaultToolkit().getScreenResolution();

```

**Review by SJuliez — CHANGES_REQUESTED** [review]
Just a few formatting things and a suggestion. I think the two codechecker things are valid as well.

File: megamek/src/megamek/client/ui/swing/util/UIUtil.java (hunk 7 above)
Comment (pakfront) [reply megamek/src/megamek/client/ui/swing/util/UIUtil.java:576]: I'll leave it here for now.

**Review by SJuliez — APPROVED** [review]

**Pull request merged**